File: ocaml/21-30/22.ml
(* Create a list containing all integers within a given range. (easy)

If first argument is greater than second, produce a list in decreasing order.
*)

let rec range from _to = 
  if from = _to then [_to]
  else if from > _to then from::(range (from - 1) _to)
  else from::(range (from + 1) _to)


let _ = assert ((range 4 9) = [4; 5; 6; 7; 8; 9])
let _ = assert ((range 9 4) = [9; 8; 7; 6; 5; 4])

File: prolog/21-30/22.pl
/*
 Create a list containing all integers within a given range.
    Example:
    ?- range(4,9,L).
    L = [4,5,6,7,8,9]
*/

range(X, X, Acc, Xs) :- reverse([X|Acc], Xs).
range(Lo, Hi, Acc, Xs) :- 
  Lo < Hi,
  L1 is Lo + 1,
  range(L1, Hi, [Lo|Acc], Xs).

range(X, X, [X]).
range(Lo, Hi, Xs) :- range(Lo, Hi, [], Xs).
